lib/improver/optical_flow/optical_flow.py

```diff
@@ -0,0 +1,295 @@
+        """
+        Performs a dimensionless grid-based extrapolation of spatial data
+        using advection velocities via a backwards method.
+
+        Args:
+            data (numpy.ndarray):
+                2D numpy data array to be advected
+            grid_vel_x (numpy.ndarray):
+                Velocity in the x direction (in grid points per second)
+            grid_vel_y (numpy.ndarray):
+                Velocity in the y direction (in grid points per second)
+            timestep (int):
+                Advection time step in seconds
+            bgd (float):
+                Default output value for spatial points where data cannot be
+                extrapolated (source is out of bounds)
+
+        Returns:
+            adv_field (numpy.ndarray):
+                2D float array of advected data values
+        """
+
+        # Initialise advected field with "background" default value
+        adv_field = np.full(data.shape, bgd)
+
+        # Set up grids of data coordinates (meshgrid inverts coordinate order)
+        ydim, xdim = data.shape
+        (xgrid, ygrid) = np.meshgrid(np.arange(xdim),
+                                     np.arange(ydim))
+
+        # For each grid point on the output field, trace its (x,y) "source"
+        # location backwards using advection velocities.  The source location
+        # is generally fractional: eg with advection velocities of 0.5 grid
+        # squares per second, the value at [2, 2] is represented by the value
+        # that was at [1.5, 1.5] 1 second ago.
+        xsrc_point_frac = -grid_vel_x * timestep + xgrid.astype(float)
+        ysrc_point_frac = -grid_vel_y * timestep + ygrid.astype(float)
+
+        # For all the points where fractional source coordinates are within
+        # the bounds of the field, set the output field to 0
+        def point_in_bounds(x, y, nx, ny):
+            """Check point (y, x) lies within defined bounds"""
+            return (x >= 0.) & (x < nx) & (y >= 0.) & (y < ny)
+
+        cond1 = point_in_bounds(xsrc_point_frac, ysrc_point_frac, xdim, ydim)
+        adv_field[cond1] = 0
+
+        # Find the integer points surrounding the fractional source coordinates
+        # and check they are in bounds
+        xsrc_point_lower = xsrc_point_frac.astype(int)
+        xsrc_point_upper = xsrc_point_lower + 1
+        ysrc_point_lower = ysrc_point_frac.astype(int)
+        ysrc_point_upper = ysrc_point_lower + 1
+
+        cond2 = point_in_bounds(xsrc_point_lower, ysrc_point_lower,
+                                xdim, ydim) & cond1
+        cond3 = point_in_bounds(xsrc_point_lower, ysrc_point_upper,
+                                xdim, ydim) & cond1
+        cond4 = point_in_bounds(xsrc_point_upper, ysrc_point_lower,
+                                xdim, ydim) & cond1
+        cond5 = point_in_bounds(xsrc_point_upper, ysrc_point_upper,
+                                xdim, ydim) & cond1
```

Comment:
I think these conditions are mismatched with the calls to increment_output_array below. Try adding a unit test(s) using reversed advection velocities to prove this.